### PR TITLE
[Feature] Official implementation of SETR

### PR DESCRIPTION
**Pull request by sixiaozheng**
This is from the **official implementation** of SETR

### CODE REVIEW

**Review by sennnnn — COMMENTED** [review]
We change part of the official codes to let this implementation more standard. Then we re-training related models and testing their precision.

## cityscapes

| Method | Crop Size | Batch Size | Iteration | Set | mIoU | 
| :--: | :--: | :--: | :--: | :--: | :--: |
| SETR-Naive | 768x768 | 8 | 40000 | val | 77.29 |
| SETR-Naive | 768x768 | 8 | 80000 | val | - |
| SETR-MLA | 768x768 | 8 | 40000 | val | 75.12 |
| SETR-MLA | 768x768 | 8 | 80000 | val | - |
| SETR-PUP | 768x768 | 8 | 40000 | val | 78.27 |
| SETR-PUP | 768x768 | 8 | 80000 | val | - |
| SETR-Naive-Deit | 768x768 | 8 | 40000 | val | 77.85 |
| SETR-Naive-Deit | 768x768 | 8 | 80000 | val | - |
| SETR-MLA-Deit | 768x768 | 8 | 40000 | val | 76.29 |
| SETR-MLA-Deit | 768x768 | 8 | 80000 | val | - |
| SETR-PUP-Deit | 768x768 | 8 | 40000 | val | - |
| SETR-PUP-Deit | 768x768 | 8 | 80000 | val | - |

## ade

| Method | Crop Size | Batch Size | Iteration | Set | mIoU | 
| :--: | :--: | :--: | :--: | :--: | :--: |
| SETR-Naive | 512x512 | 16 | 160000 | val | - |
| SETR-MLA | 512x512 | 8 | 160000 | val | 47.52 |
| SETR-MLA | 512x512 | 16 | 160000 | val | 48.16 |
| SETR-PUP | 512x512 | 16 | 160000 | val | 48.27 |

## pascal context

| Method | Crop Size | Batch Size | Iteration | Set | mIoU | 
| :--: | :--: | :--: | :--: | :--: | :--: |
| SETR-Naive | 480x480 | 16 | 80000 | val | - |
| SETR-MLA | 480x480 | 8 | 80000 | val | - |
| SETR-MLA | 480x480 | 16 | 80000 | val | - |
| SETR-PUP | 480x480 | 16 | 80000 | val | - |

File: configs/setr/setr_mla_480x480_80k_pascal_context_bs_8.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    decode_head=dict(img_size=(480, 480), num_classes=60),
+    auxiliary_head=[aux_alpha, aux_beta, aux_gamma, aux_delta],
+    test_cfg=dict(mode='slide', crop_size=(480, 480), stride=(320, 320)))
+
+optimizer = dict(
+    lr=0.001,
+    weight_decay=0.0,
+    paramwise_cfg=dict(custom_keys={'head': dict(lr_mult=10.)}))
+
+find_unused_parameters = False
```
Comment (xvjiarui) [inline configs/setr/setr_mla_480x480_80k_pascal_context_bs_8.py:61]: This is `False` by default. We may remove it.

File: mmseg/datasets/ade.py [hunk 2]
```diff
@@ -82,3 +89,81 @@ def __init__(self, **kwargs):
             seg_map_suffix='.png',
             reduce_zero_label=True,
             **kwargs)
+
+    def results2img(self, results, imgfile_prefix, to_label_id):
```
Comment (xvjiarui) [inline mmseg/datasets/ade.py:93]: We may make this a separate PR.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            else:
+                raise NotImplementedError
+        elif self.num_convs == 4:
+            if self.num_up_layer == 4:
+                x = self.conv_0(x)
+                x = self.unified_bn_fc_0(x)
+                x = F.relu(x, inplace=True)
+                x = F.interpolate(
+                    x,
+                    size=x.shape[-1] * 2,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:140]: Incorrect for non-square input.

File: configs/_base_/datasets/cityscapes_768x768.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+_base_ = './cityscapes.py'
```
[the remark this answers is not in the document]
Comment (sennnnn) [reply configs/_base_/datasets/cityscapes_768x768.py:1]: Already add

File: configs/_base_/models/setr_mla.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    type='SETRMLAAUXHead',
+    in_channels=256,
+    channels=512,
+    in_index=3,
+    img_size=(768, 768),
+    mla_channels=256,
+    num_classes=19,
+    align_corners=False,
+    loss_decode=dict(
+        type='CrossEntropyLoss', use_sigmoid=False, loss_weight=0.4))
```
[the remark this answers is not in the document]
Comment (sennnnn) [reply configs/_base_/models/setr_mla.py]: Already modify.

File: configs/_base_/models/setr_mla.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+model = dict(
+    type='EncoderDecoder',
+    pretrained='https://github.com/rwightman/pytorch-image-models/releases/\
+download/v0.1-vitjx/jx_vit_large_p16_384-b3be5167.pth',
```
[the remark this answers is not in the document]
Comment (sennnnn) [reply configs/_base_/models/setr_mla.py]: Already modify.

File: mmseg/models/decode_heads/setr_mla_head.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import build_norm_layer
+
+from ..builder import HEADS
+from .decode_head import BaseDecodeHead
+
+
+class MLAModule(nn.Module):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_mla_head.py:10]: Refer to PPM in PSPHead.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                 norm_layer=dict(type='LN', eps=1e-6, requires_grad=True),
+                 num_convs=2,
+                 up_mode='bilinear',
+                 num_up_layer=1,
+                 kernel_size=3,
+                 **kwargs):
+
+        assert num_convs in [
+            2, 4
+        ], 'naive head (num_convs = 2), pup head (num_convs=4).'
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:38]: Reformat.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+        if self.num_convs == 2:
+            self.conv_0 = nn.Conv2d(
+                embed_dim,
+                256,
+                kernel_size=kernel_size,
+                stride=1,
+                padding=int(kernel_size - 1) // 2)
+            self.conv_seg = nn.Conv2d(
+                256, out_channel, kernel_size=1, stride=1)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:63]: Is it necessary to override it?

File: mmseg/models/decode_heads/setr_up_head.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+                256, 256, kernel_size=kernel_size, stride=1, padding=1)
+            self.conv_3 = nn.Conv2d(
+                256, 256, kernel_size=kernel_size, stride=1, padding=1)
+            self.conv_seg = nn.Conv2d(
+                256, out_channel, kernel_size=1, stride=1)
+
+            _, self.unified_bn_fc_0 = build_norm_layer(self.norm_cfg, 256)
+            _, self.unified_bn_fc_1 = build_norm_layer(self.norm_cfg, 256)
+            _, self.unified_bn_fc_2 = build_norm_layer(self.norm_cfg, 256)
+            _, self.unified_bn_fc_3 = build_norm_layer(self.norm_cfg, 256)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:81]: Could we make it more generalizable? Just like PSPNet.

File: mmseg/models/necks/mla_neck.py [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
+        self.mla_p5 = ConvModule(
+            mla_channels,
+            mla_channels,
+            kernel_size=3,
+            bias=False,
+            padding=1,
+            norm_cfg=norm_cfg,
+            act_cfg=act_cfg)
```
Comment (xvjiarui) [inline mmseg/models/necks/mla_neck.py:74]: Too long and too much duplicate.

File: mmseg/models/decode_heads/setr_mla_head.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+            Default: [256, 256, 256, 256].
+        mla_channels (int): MLA module output channels. Default: 128.
+        norm_cfg (dict|None): Config of norm layers. Default: None.
+        act_cfg (dict): Config of activation layers.
+            Default: dict(type='ReLU')
+    """
+
+    def __init__(self,
+                 align_corners,
+                 aggregation_stages,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_mla_head.py:26]: This number is the same as `len(in_channels)`, so we could omit this arg.

File: mmseg/models/decode_heads/setr_mla_head.py [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+                 in_channels=[256, 256, 256, 256],
+                 mla_channels=128,
+                 norm_cfg=None,
+                 act_cfg=dict(type='ReLU')):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_mla_head.py:30]: Make all args without default. Leave the default values to `SETRMLAHead`.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+                 in_channels,
+                 out_channels=256,
+                 num_convs=1,
+                 up_mode='bilinear',
+                 up_scale=4,
+                 align_corners=False,
+                 num_up_layer=0,
+                 kernel_size=3,
+                 norm_cfg=None,
+                 act_cfg=dict(type='ReLU')):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:22]: Leave all default values to `SETRUPHead`.

File: mmseg/models/decode_heads/setr_mla_head.py (hunk 12 above)
Comment (xvjiarui) [reply mmseg/models/decode_heads/setr_mla_head.py:26]: We may also accept integer input.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import ConvModule, build_norm_layer
+
+from ..builder import HEADS
+from ..utils import trunc_normal_
+from .decode_head import BaseDecodeHead
+
+
+class UPModule(nn.ModuleList):
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:10]: Missing docstring.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __init__(self,
+                 in_channels,
+                 out_channels=256,
+                 num_convs=1,
+                 up_mode='bilinear',
+                 up_scale=4,
+                 align_corners=False,
+                 num_up_layer=0,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:19]: Why `num_up_layer != num_convs` ?

File: mmseg/models/decode_heads/setr_up_head.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+from .decode_head import BaseDecodeHead
+
+
+class UPModule(nn.ModuleList):
+
+    def __init__(self,
+                 in_channels,
+                 out_channels=256,
+                 num_convs=1,
+                 up_mode='bilinear',
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:16]: We may not make this an arg. Since we always prefer `bilinear`.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        self.num_up_layer = num_up_layer
+        self.kernel_size = kernel_size
+
+        for i in range(num_convs):
+            if i == 0:
+                tmp_in_channels = in_channels
+                tmp_out_channels = out_channels
+            else:
+                tmp_in_channels = out_channels
+                tmp_out_channels = out_channels
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:37]: These lines are not necessary. 

We just need 1 variable `last_out_channels` to record the output channels out last conv.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+            self.append(
+                ConvModule(
+                    in_channels=tmp_in_channels,
+                    out_channels=tmp_out_channels,
+                    kernel_size=kernel_size,
+                    stride=1,
+                    padding=int(kernel_size - 1) // 2,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:44]:
```suggestion
                    padding=(kernel_size - 1) // 2,
```

File: mmseg/models/decode_heads/setr_up_head.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        up_scale (int): The scale factor of interpolate. Default:4.
+        num_up_layer (str): Nunber of upsampling layers. Default: 1.
+        kernel_size (int): The kernel size of convolution when decoding
+            feature information from backbone. Default: 3.
+    """
+
+    def __init__(self,
+                 norm_layer=dict(type='LN', eps=1e-6, requires_grad=True),
+                 num_convs=1,
+                 up_mode='bilinear',
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:86]: We may omit this.

File: mmseg/models/necks/mla_neck.py [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+                    kernel_size=3,
+                    padding=1,
+                    norm_cfg=norm_cfg,
+                    act_cfg=act_cfg))
+
+    def forward(self, inputs):
+
+        # feat_list -> [p2, p3, p4, p5]
+        feat_list = []
+        for x, op in zip(inputs, self.channel_proj):
```
Comment (xvjiarui) [inline mmseg/models/necks/mla_neck.py:39]:
```suggestion
        for x, conv in zip(inputs, self.channel_proj):
```

File: mmseg/models/decode_heads/setr_up_head.py [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+    def forward(self, x):
+        # There is an upsample defined in segmentor class..
+        up_rest_count = self.num_up_layer
+        for op in self:
+            x = op(x)
+            if up_rest_count > 0:
+                x = F.interpolate(
+                    x,
+                    size=(x.shape[-2] * self.up_scale,
+                          x.shape[-1] * self.up_scale),
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:57]: For fixed size upsample, we may use `nn.Upsample`.

File: mmseg/models/decode_heads/setr_up_head.py [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+        assert kernel_size in [1, 3], 'kernel_size must be 1 or 3.'
+
+        super(SETRUPHead, self).__init__(**kwargs)
+
+        assert isinstance(self.in_channels, int)
+
+        _, self.norm = build_norm_layer(norm_layer, self.in_channels)
+
+        self.up = nn.ModuleList()
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:40]:
```suggestion
        self.conv_up = nn.ModuleList()
```

File: mmseg/models/decode_heads/setr_up_head.py [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+
+    def forward(self, x):
+        x = self._transform_inputs(x)
+
+        n, c, h, w = x.shape
+        x = x.reshape(n, c, h * w).transpose(2, 1)
+        x = self.norm(x)
+        x = x.transpose(1, 2).reshape(n, c, h, w)
+
+        for op in self.up:
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_up_head.py:78]:
```suggestion
        for up_conv in self.up_convs:
```

File: mmseg/models/decode_heads/setr_mla_head.py [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+                        act_cfg=self.act_cfg),
+                    ConvModule(
+                        in_channels=mla_channels,
+                        out_channels=mla_channels,
+                        kernel_size=3,
+                        padding=1,
+                        norm_cfg=self.norm_cfg,
+                        act_cfg=self.act_cfg),
+                    nn.Upsample(
+                        scale_factor=4,
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_mla_head.py:49]: Make this an argument.

File: mmseg/models/decode_heads/setr_mla_head.py [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+        self.mla_channels = mla_channels
+
+        num_inputs = len(self.in_channels)
+
+        # Refer to self.cls_seg settings of BaseDecodeHead
+        assert self.channels == num_inputs * mla_channels
+
+        self.mlahead = nn.ModuleList()
+        for i in range(num_inputs):
+            self.mlahead.append(
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/setr_mla_head.py:32]:
```suggestion
            self.up_convs.append(
```

File: configs/setr/README.md [hunk 27]
```diff
@@ -0,0 +1,25 @@
...
+| ------ | -------- | --------- | ---------- | ------- | -------- | -------------- | ----- | ------------: | ------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| SETR-Naive | ViT | 512x512  | 16          | 160000   | -        | -              | 48.47 |             - | [config](https://github.com/open-mmlab/mmsegmentation/blob/master/configs/setr/setr_naive_512x512_160k_b16_ade20k.py)  | [model](https://download.openmmlab.com/mmsegmentation/v0.5/setr/setr_naive_512x512_160k_b16_ade20k/setr_naive_512x512_160k_b16_ade20k_20210521_155347-****.pth) &#124; [log](https://download.openmmlab.com/mmsegmentation/v0.5/setr/setr_naive_512x512_160k_b16_ade20k/setr_naive_512x512_160k_b16_ade20k_20210521_155347.log.json)     |
```
Comment (xvjiarui) [inline configs/setr/README.md:22]: We may also add missing inf time and ms+flip test results.

**Comment by Junjun2016** [comment]
Hi @sennnnn 
Please add information to the Supported backbones and Supported methods in English and Chinese README.md

File: configs/_base_/datasets/cityscapes_768x768.py [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+crop_size = (768, 768)
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='LoadAnnotations'),
+    dict(type='Resize', img_scale=(2049, 1025), ratio_range=(0.5, 2.0)),
```
Comment (Junjun2016) [inline configs/_base_/datasets/cityscapes_768x768.py:8]: img_scale=(2049, 1025) --》img_scale=(2048, 1024)

File: configs/_base_/datasets/cityscapes_768x768.py [hunk 29]
```diff
@@ -0,0 +1,35 @@
...
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='Pad', size=crop_size, pad_val=0, seg_pad_val=255),
+    dict(type='DefaultFormatBundle'),
+    dict(type='Collect', keys=['img', 'gt_semantic_seg']),
+]
+test_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(
+        type='MultiScaleFlipAug',
+        img_scale=(2049, 1025),
```
Comment (Junjun2016) [inline configs/_base_/datasets/cityscapes_768x768.py:21]: img_scale=(2049, 1025) --》img_scale=(2048, 1024)

**Review by xvjiarui — APPROVED** [review]

File: mmseg/models/decode_heads/setr_up_head.py [hunk 30]
```diff
@@ -0,0 +1,75 @@
...
+                        padding=int(kernel_size - 1) // 2,
+                        norm_cfg=self.norm_cfg,
+                        act_cfg=self.act_cfg),
+                    nn.Upsample(
+                        scale_factor=up_scale,
+                        mode='bilinear',
+                        align_corners=self.align_corners)))
+            in_channels = out_channels
+
+    def init_weights(self):
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/setr_up_head.py:58]: Use `init_cfg` to control it?
